docs/design/proposals/rbd-volume-healer.md

```diff
@@ -0,0 +1,94 @@
+rbd driver to mount the rbd images onto the application node. Here on
+at cephcsi we will also have a userspace way of mounting the RBD images,
+via RBD-NBD.
+
+[Rbd-nbd](https://docs.ceph.com/en/latest/man/8/rbd-nbd/) is a client for
+RADOS block device (rbd) images like the existing rbd kernel module. It
+will map an rbd image to an NBD (Network Block Device) device, allowing
+access to it as a regular local block device.
+
+![csi-rbdnbd](https://user-images.githubusercontent.com/12432241/125400940-23e75d00-e3d0-11eb-806c-bab09318de54.png)
+
+It’s worth making a note that the rbd-nbd processes will run on the
+client-side, which is inside the `csi-rbdplugin` node plugin.
+
+### Advantages of userspace mounters
+
+- Adding changes to client logic to accommodate new features will be easy
+- Container upgrades will be independent of the host node, which means if
+  there any changes in the client logic, we don’t have to reboot the node
+  as the changes will be shipped inside the container.
+- We will be a better citizen in K8s by switching to the userspace model.
+- Unlike krbd, rbd-nbd uses librbd user-space library that gets most of the
+  development focus, and hence rbd-nbd will be future rich.
+- Unlike rbd-nbd, krbd exhibits additional drawback, being entirely kernel
+  space impacts fault-tolerance as any kernel panic affects a whole node
+  not only a single Pod using RBD storage. Thanks to the rbd-nbd’s
+  userspace design, this doesn’t bother here.
+
+### Side effects of userspace mounters
+
+Since the rbd-nbd processes run per volume map on the client side i.e.
+inside the `csi-rbdplugin` node plugin, a restart of the node plugin will
+terminate all the rbd-nbd processes, and there is no way to restore
+these processes back to life currently, which could lead to IO errors
+on all the application pods.
+
+![nodeplugin-restart](https://user-images.githubusercontent.com/12432241/125401399-ba1b8300-e3d0-11eb-9db2-e00f11a96af8.png)
+
+This is where the volume Healer could help.
+
+## Volume Healer
+
+Volume healer runs on the start of rbd node-plugin and runs within the
+nodeplugin driver context.
+
+Volume healer does the below,
+
+- Get the Volume attachment list for the current node where it is running
+- Filter the volume attachments list through matching driver name and
+  status attached
+- For each volume attachment get the respective PV information and check
+  the criteria of PV Bound, mounter type
+- Build the StagingPath where rbd images PVC is mounted, based on the
+  KUBELET path and PV object
+- Construct the NodeStageVolume() request and send Request to CSI Driver.
+- The NodeStageVolume() has a way to identify calls received from the
+  healer and when executed from the healer context, it just runs in the
+  minimal required form, where it fetches the previously mapped device to
+  the image, and the respective secrets and finally ensures to bringup the
+  respective process back to life. Thus enabling IO to continue.
+
```

Comment:
No amount of "healing" is going to fix potential data loss that could happen because application simply failed to write data to underlying mount point. Are we sure this is what we want?

Reply:
Not to mention, this would make driver very much kube specific. kinda defeats the purpose of CSI, no?

Reply:
+1, this feels very error prone. I like the kernel implementation, because it's rock stable.

Reply:
Using userspace clients for storage is definitely more complex than using the kernel equivalents. However, we would like to persuade this feature, as there are benefits as well. The current available feature (since the beginning of ceph-csi) has been labelled alpha. There are many improvements needed in both Ceph-CSI as well as in the Ceph rbd-nbd component. @pkalever is working on both, and https://github.com/ceph/ceph-csi/projects/13 contains the tasks that still are pending.

The main painpoint users can be affected with, is the cancelled I/O while the block-device is temporarily unavailable. This will cause filesystems to abort and remount themselves as read-only. There is no practical recovery from such a state. To prevent this, there are discussions with the Ceph team to run `fsfreeze` before detaching the userspace process, and unfreezing the filesystem afterwards. How this can be done for non-filesystem users remains to be seen (device-mapper can do it, maybe there is a generic syscall/ioctl). @pkalever, please create an issue so that we can track the Ceph discussions (add URLs to mails/... there).

Even with this volume healer, ceph-csi will keep the rbd-nbd feature in Alpha state. But, we do hope to get a little closer into gaining confidence to run real workloads on top of it.